File: tests/correct/helper/treiber-stack-dynamic/variants/main0.c
#include <stdlib.h>
#include <stdbool.h>
#include <pthread.h>
#include <assert.h>

#include "../main.c"

int main()
{
	int i = 0;

	num_threads = readers + writers + rdwr;

	atomic_init(&x[1], 0);
	atomic_init(&x[2], 0);

	init_stack(&stack, num_threads);
	for (int j = 0; j < num_threads; j++)
		param[j] = j;
	for (int j = 0; j < writers; j++, i++)
		pthread_create(&threads[i], NULL, threadW, &param[i]);
	for (int j = 0; j < readers; j++, i++)
		pthread_create(&threads[i], NULL, threadR, &param[i]);
	for (int j = 0; j < rdwr; j++, i++)
		pthread_create(&threads[i], NULL, threadRW, &param[i]);

	for (i = 0; i < num_threads; i++)
		pthread_join(threads[i], NULL);

	/* bool correct = false; */
	//correct |= (a == 17 || a == 37 || a == 0);
	//MODEL_ASSERT(correct);

	/* free(param); */
	/* free(threads); */
	/* free(stack); */

	return 0;
}

File: rust/tests/correct/data-structures/treiber-stack-dynamic/variants/main0.rs
#![allow(unused_imports)]

use std::{mem::MaybeUninit, sync::atomic::AtomicI32};
use std::thread::JoinHandle;

#[path = "../common.rs"]
mod common;
use common::{threadW, threadR, threadRW, NUM_THREADS, PARAM, WRITERS, READERS, RDWR, X, threadW0, threadW1, threadR2};
use common::my_stack::MAX_THREADS;


pub fn main() {
    let mut i = 0;

    let mut threads: [MaybeUninit<JoinHandle<()>>; MAX_THREADS] = std::array::from_fn(|_| MaybeUninit::uninit());

    //num_threads: Done statically

    //X[1].write(AtomicI32::new(0)); //Not needed => Static initialization
    //X[2].write(AtomicI32::new(0));

    //init stack: Done statically
    for j in 0..NUM_THREADS {
        unsafe {PARAM[j] = j};
    }
    for _ in 0..WRITERS {
        let pj = unsafe {PARAM[i]};
        threads[i].write(std::thread::spawn(move || {threadW(pj)}));
        i += 1;
    }
    for _ in 0..READERS {
        let pj = unsafe {PARAM[i]};
        threads[i].write(std::thread::spawn(move || {threadR(pj)}));
        i += 1;
    }
    for _ in 0..RDWR {
        let pj = unsafe {PARAM[i]};
        threads[i].write(std::thread::spawn(move || {threadRW(pj)}));
        i += 1;
    }

    for j in 0..NUM_THREADS {
        unsafe { threads[j].assume_init_mut().join_keep().unwrap() };
    }

    //let h1 = std::thread::spawn_f(threadW0);
    //let h2 = std::thread::spawn_f(threadW1);
    //let h3 = std::thread::spawn_f(threadR2);
    /*let h1 = std::thread::spawn_f_args(threadW, 0usize);
    let h2 = std::thread::spawn_f_args(threadW, 1usize);
    let h3 = std::thread::spawn_f_args(threadR, 2usize);
    h1.join().unwrap();
    h2.join().unwrap();
    h3.join().unwrap();*/

    /*std::thread::scope(|s| {
        s.spawn(|| {threadW(0usize)});
        s.spawn(|| {threadR(1usize)});
        //s.spawn(|| {threadR(2usize)});
        //let mut i = 0;
        /*for j in 0..WRITERS {
            let pj = unsafe {PARAM[i]};
            s.spawn(move || {threadW(pj)});
            i += 1;
        }
        for j in 0..READERS {
            let pj = unsafe {PARAM[i]};
            s.spawn(move || {threadR(pj)});
            i += 1;
        }*/
        /*for j in 0..RDWR {
            let pj = unsafe {PARAM[i]};
            s.spawn(move || {threadRW(pj)});
            i += 1;
        }*/
        6
    });*/
}